source/tutorials/module-system/module-system.md

```diff
@@ -0,0 +1,1289 @@
+
+### What do you need?
+
+- Familiarity with data types and general programming concepts
+- A {ref}`Nix installation <install-nix>` to run the examples
+- Intermediate proficiency in reading and writing the Nix language
+
+You will use two helper scripts for this exercise.
+Download {download}`map <files/map>` and {download}`geocode <files/geocode>` to your working directory.
+
+:::{warning}
+To run the examples in this tutorial, you will need a [Google API key](https://developers.google.com/maps/documentation/maps-static/start#before-you-begin) in `$XDG_DATA_HOME/google-api/key`.
+:::
+
+### How long will it take?
+
+This is a very long tutorial.
+Prepare for at least 3 hours of work.
+
+## Empty module
+
+We have to start somewhere.
+The simplest module is just a function that takes any attributes and returns an empty attribute set.
+
+Write the following into a file called `default.nix`:
+
+```nix
+# default.nix
+{ ... }:
+{
+
+}
+```
+
+## Module Arguments
+
+We will need some helper functions, which will come from the Nixpkgs library.
+Start by changing the first line in `default.nix`:
+
+```diff
+# default.nix
+- { ... }:
++ { lib, ... }:
+{
+
+}
+```
+
+Now the module is a function which takes *at least* one argument, called `lib`, and may accept other arguments (expressed by the ellipsis `...`).
+
+This will make Nixpkgs library functions available within the function body.
+The `lib` argument is passed automatically by the module system.
+
+:::{note}
+The ellipsis `...` is necessary because arbitrary arguments can be passed to modules.
+:::
+
+## Declaring Options
+
+To set any values, the module system first has to know which ones are allowed.
+
+This is done by declaring *options* that specify which values can be set and used elsewhere.
+Options are declared by adding an attribute under the top-level `options` attribute, using `lib.mkOption`.
+
+In this section, you will define the `scripts.output` option.
+
+Change `default.nix` to include the following declaration:
+
+```diff
+# default.nix
+ { lib, ... }: {
+
++ options = {
++   scripts.output = lib.mkOption {
++     type = lib.types.lines;
++   };
++ };
+
+ }
+```
+
+While many attributes for customizing options are available, the most important one is `type`, which specifies which values are valid for an option.
+There are several types available under [`lib.types`](https://nixos.org/manual/nixos/stable/#sec-option-types-basic) in the Nixpkgs library.
+
+You have just declared `scripts.output` with the `lines` type, which specifies that the only valid values are strings, and that multiple definitions should be joined with newlines.
+
+:::{note}
+The name and attribute path of the option is arbitrary.
+Here we use `scripts`, because we will add another script later, and call this one `output`, because it will output the resulting map.
+:::
+
+## Evaluating modules
```

Comment:
This section should be at the very start. `{ }` is not a module, it's just an attribute set. Only calling `evalModules` on it makes it be interpreted as module. So it's weird to talk about modules before `evalModules` is introduced.